icicle_v3/backend/cpu/include/tasks_manager.h

```diff
@@ -0,0 +1,309 @@
+  inline bool has_result();
+  inline bool is_idle();
+  inline void set_working();
+  inline void set_pending_result();
+  inline void set_handled_result();
+
+  /**
+   * @brief link this task with the tail idx of the Worker it belongs to. Used only once after construction.
+   */
+  inline void link_father_tail(int* tail_pointer) { father_fifo_tail = tail_pointer; } 
+
+  /**
+   * @brief wait for a specific task to finish executing. This is a blocking function.
+   */
+  void wait_done(); // NOTE the user must handle this task's result before asking for new free tasks  
+
+private:
+  /**
+   * @enum containing the valid states of a task.
+   */
+  enum eTaskStatus {IDLE, DISPATCHED, WORKING, PENDING_RESULT}; // TODO CAPS and eTaskStatus
+  std::atomic<eTaskStatus> status; // current task state. Atomic to ensure proper sync between main and worker threads.
+  int* father_fifo_tail; // pointer to father's tail.
+};
+
+
+bool TaskBase::is_ready_for_work() {
+  return status.load(std::memory_order_acquire) == DISPATCHED;
+}
+
+bool TaskBase::can_push_task() {
+  eTaskStatus curr_status = status.load(std::memory_order_acquire);
+  return curr_status == PENDING_RESULT || curr_status == IDLE;
+}
+
+bool TaskBase::has_result() {
+  return status.load(std::memory_order_acquire) == PENDING_RESULT;
+}
+
+bool TaskBase::is_idle() {
+  return status.load(std::memory_order_acquire) == IDLE;
+}
+
+void TaskBase::dispatch() {
+  (*father_fifo_tail)++;
+  status.store(DISPATCHED, std::memory_order_release);
+}
+
+void TaskBase::set_working() {
+  status.store(WORKING, std::memory_order_release);
+}
+
+void TaskBase::set_pending_result() {
+  status.store(PENDING_RESULT, std::memory_order_release);
+}
+
+void TaskBase::set_handled_result() {
+  status.store(IDLE, std::memory_order_release);
+}
+
+void TaskBase::wait_done() {
+  while (!has_result()) {
+    std::this_thread::sleep_for(std::chrono::microseconds(MANAGER_SLEEP_USEC));
+  }
+}
+
+template<class Task>
+TasksManager<Task>::Worker::Worker() 
+: mTasksFifo(TASKS_PER_THREAD), 
+  tail(0),
+  head(0),
+  kill(false)
+  {
+    // Tail idx linking can only be done after task and worker initialization
+    for (Task& task : mTasksFifo) task.link_father_tail(&tail);
+    // Init thread only after finishing all other setup to avoid data races
+    task_executor = std::thread(&TasksManager<Task>::Worker::run, this);
+  }
+
+template<class Task>
+TasksManager<Task>::Worker::~Worker() {
+  kill = true;
+  task_executor.join();
+}
+
+template<class Task>
+void TasksManager<Task>::Worker::run() {
+  while (true) {
+    for (head = 0; head < mTasksFifo.size(); head++)
+    {      
+      Task* task = &mTasksFifo[head];
+      if (!task->is_ready_for_work()) {
+        // Sleep as the thread apparently isn't fully utilized currently
+        std::this_thread::sleep_for(std::chrono::microseconds(THREAD_SLEEP_USEC));       
+        continue;
+      }
+      task->set_working();
+      task->execute();
```

Comment:
have the execute call set_working(). It's too easy to mess it up like that.

Reply:
the issue is that execute is a virtual function will be defined by the user
so it (and set_pending_result) is called in run to avoid cases where the user forgets to update the status and the code is stuck